File: scripts/js/validMoves.js
import { whitePositions, blackPositions, deleteFromMap } from "./main.js";

// up right down left
let rookDirections  = [[0,-1], [1,0], [0,1], [-1,0]];
// up-right bottom-right bottom-left up-left
let bishopDirections = [[1,-1], [1,1], [-1,1], [-1,-1]];
// clock wise starting from up 
let knightDirections = [[1,-2], [2,-1], [2,1], [1,2], [-1,2], [-2,1], [-2,-1], [-1,-2]];
let kingDirections   = [[0,-1], [1,0], [0,1], [-1,0], [1,-1], [1,1], [-1,1], [-1,-1]];

export function getValidMovesWhite(board) {

    let moves = [];
    whitePositions.forEach((value, key) => {
        let a = getValidMoves(board, "w"+value, key[0], key[1]);
        moves.push.apply(moves, validMovesFromArray(board, key[0], key[1], a));
        // a.forEach(move => {
        //     let temp = board[move[1]][move[0]];
        //     // start
        //     board[key[1]][key[0]] = null;
        //     // target
        //     board[move[1]][move[0]] = "w"+value;
        //     if (value == 'k') kingPos = [move[0][move[1]]];

        //     if (!isInCheck(board, 'w', kingPos[0], kingPos[1])) {
        //         if (!moves.some(a => a[0] == move[0] && a[1] == move[1])) moves.push(move);
        //     }

        //     // start
        //     board[key[1]][key[0]]   = "w"+value;
        //     if (value == 'k') kingPos = [key[0][key[1]]];
        //     // end
        //     board[move[1]][move[0]] = temp;    
        // });
    });
    return moves;
}

export function getValidMovesBlack(board) {

    let moves = [];
    blackPositions.forEach((value, key) => {
        let a = getValidMoves(board, "b"+value, key[0], key[1]);
        moves.push.apply(moves, validMovesFromArray(board, key[0], key[1], a));
        // a.forEach(move => {
        //     let temp = board[move[1]][move[0]];
        //     // start
        //     board[key[1]][key[0]]   = null;
        //     // target
        //     board[move[1]][move[0]] = "b"+value;
        //     if (value == 'k') kingPos = [move[0][move[1]]];

        //     if (!isInCheck(board, 'b')) {
        //         if (!moves.some(a => a[0] == move[0] && a[1] == move[1])) moves.push(move);
        //     }
            
        //     // start
        //     board[key[1]][key[0]]   = "b"+value;
        //     if (value == 'k') kingPos = [key[0][key[1]]];
        //     // end
        //     board[move[1]][move[0]] = temp;    
        // });
    });
    return moves;
}

export function validMovesFromArray(board, startX, startY, array) {
    
    let color = board[startY][startX].charAt(0);
    let piece = board[startY][startX].charAt(1);
    let positions = color === "w" ? whitePositions : blackPositions;
    let kingPos = [];
    positions.forEach((value, key) => {
        if (value == 'k') kingPos = key;
    });
    let moves = [];

    array.forEach(move => {
        let temp = board[move[1]][move[0]];
        // start
        board[startY][startX]   = null;
        // target
        board[move[1]][move[0]] = color+piece;

        if (piece == 'k') kingPos = [move[0], move[1]];

        if (!isInCheck(board, color, kingPos)) {
            if (!moves.some(a => a[0] == move[0] && a[1] == move[1])) moves.push(move);
        }
        
        // start
        board[startY][startX] = color+piece;
        if (piece == 'k') kingPos = [startY, startX];
        // end
        board[move[1]][move[0]] = temp;   

        // if (piece == 'k') {
        //     if (color == 'w') {
        //         // whitePositions.set([move[0], move[1]], color);
        //         // deleteFromMap(whitePositions, [move[0], move[1]]);
        //         // if (temp) blackPositions.set([move[0], move[1]], temp.charAt(1));
        //         deleteFromMap(whitePositions, [move[0], move[1]]);
        //         whitePositions.set([startX, startY], piece);
        //         if (temp) blackPositions.set([move[0], move[1]], temp.charAt(1));
        //     } else {
        //         deleteFromMap(blackPositions, [move[0], move[1]]);
        //         blackPositions.set([startX, startY], piece);
        //         if (temp) whitePositions.set([move[0], move[1]], temp.charAt(1));
        //     }
        // }
    });
    return moves;
}


export function getValidMoves (board, piece, xPos, yPos) {

    const pieceType = piece.substring(1); // e.g., 'p' for pawn, 'r' for rook, etc.
    let moves = [];
    // need to account for taking pieces
    switch (pieceType) {
        case 'p': // Pawn move (basic example)
            moves = validMovesFromArray(board, xPos, yPos, pawnMoves(board, piece[0], xPos, yPos));
            break;
        case 'r': // Rook move
            moves = validMovesFromArray(board, xPos, yPos, rookMoves(board, piece[0], xPos, yPos));
            break;
        case 'n': // Knight move
            moves = validMovesFromArray(board, xPos, yPos, knightMoves(board, piece[0], xPos, yPos));
            break;
        case 'b': // Bishop move
            moves = validMovesFromArray(board, xPos, yPos, bishopMoves(board, piece[0], xPos, yPos));
            break;
        case 'q': // Queen move
            moves = validMovesFromArray(board, xPos, yPos, queenMoves(board, piece[0], xPos, yPos));
            break;
        case 'k': // King move
            moves = validMovesFromArray(board, xPos, yPos, kingMoves(board, piece[0], xPos, yPos));
            break;
    }

    return moves;
}

export function checkColor(piece, color) {
    if (piece === null) return false;
    return piece[0] === color;
}

export function checkPiece(piece1, piece2) {
    if (piece1 === null) return false;
    return piece1[1] === piece2;
}

// no en pessant yet
function pawnMoves(board, color, xPos, yPos) {

    let moves = [];
    let oppColor = (color === 'w') ? 'b' : 'w';
    // can eat of the opposite color
   
    if (yPos-1 >= 0) {
        if (xPos+1  < 8 && checkColor(board[yPos-1][xPos+1], oppColor)) moves.push([xPos+1, yPos-1]);
        if (xPos-1 >= 0 && checkColor(board[yPos-1][xPos-1], oppColor)) moves.push([xPos-1, yPos-1]);
    }
    
    if (color == 'w') {
         // can eat of the opposite color
        if (yPos-1 >= 0) {
            if (xPos+1  < 8 && checkColor(board[yPos-1][xPos+1], oppColor)) moves.push([xPos+1, yPos-1]);
            if (xPos-1 >= 0 && checkColor(board[yPos-1][xPos-1], oppColor)) moves.push([xPos-1, yPos-1]);
        }
        // move twice at start
        if (yPos == 6 && board[yPos-2][xPos] === null) moves.push([xPos, yPos-2]);
        // move once forward
        if (yPos > 0  && board[yPos-1][xPos] === null) moves.push([xPos, yPos-1]);
    // if black
    } else {
        // can eat of the opposite color
        if (yPos+1 < 8) {
            if (xPos+1  < 8 && checkColor(board[yPos+1][xPos+1], oppColor)) moves.push([xPos+1, yPos+1]);
            if (xPos-1 >= 0 && checkColor(board[yPos+1][xPos-1], oppColor)) moves.push([xPos-1, yPos+1]);
        }
        // move twice at start
        if (yPos == 1 && board[yPos+2][xPos] === null) moves.push([xPos, yPos+2]);
        // move once forward
        if (yPos <  7 && board[yPos+1][xPos] === null) moves.push([xPos, yPos+1]);
    }

    return moves;
}

function rookMoves(board, color, xPos, yPos) {
    let moves = [];
    let oppColor = (color === 'w') ? 'b' : 'w';
    let xTemp;
    let yTemp;
    for (let i=0; i < rookDirections.length; i++) {
        xTemp = xPos;
        yTemp = yPos;
        while (xTemp >= 0 && xTemp < 8 && yTemp >= 0 && yTemp < 8) {
            xTemp += rookDirections[i][0];
            yTemp += rookDirections[i][1];
            if (!(xTemp >= 0 && xTemp < 8 && yTemp >= 0 && yTemp < 8)) break;
            // stop if same color
            if (checkColor(board[yTemp][xTemp], color)) break;
            moves.push([xTemp, yTemp]);
            // at first that can eat, break
            if (checkColor(board[yTemp][xTemp], oppColor)) break;
        }
    }
    return moves;
}

function knightMoves(board, color, xPos, yPos) {
    let moves = [];
    let oppColor = (color === 'w') ? 'b' : 'w';
    let xTemp;
    let yTemp;
    for (let i=0; i < knightDirections.length; i++) {
        xTemp = xPos;
        yTemp = yPos;
        if (xTemp >= 0 && xTemp < 8 && yTemp >= 0 && yTemp < 8) {
            xTemp += knightDirections[i][0];
            yTemp += knightDirections[i][1];
            if (!(xTemp >= 0 && xTemp < 8 && yTemp >= 0 && yTemp < 8)) continue;
            // stop if same color
            if (checkColor(board[yTemp][xTemp], color)) continue;
            moves.push([xTemp, yTemp]);
            // at first that can eat, continue
            if (checkColor(board[yTemp][xTemp], oppColor)) continue;
        }
    }
    return moves;
}

function bishopMoves(board, color, xPos, yPos) {
    let moves = [];
    let oppColor = (color === 'w') ? 'b' : 'w';
    let xTemp;
    let yTemp;
    for (let i=0; i < bishopDirections.length; i++) {
        xTemp = xPos;
        yTemp = yPos;
        while (xTemp >= 0 && xTemp < 8 && yTemp >= 0 && yTemp < 8) {
            xTemp += bishopDirections[i][0];
            yTemp += bishopDirections[i][1];
            if (!(xTemp >= 0 && xTemp < 8 && yTemp >= 0 && yTemp < 8)) break;
            // stop if same color
            if (checkColor(board[yTemp][xTemp], color)) break;
            moves.push([xTemp, yTemp]);
            // at first that can eat, break
            if (checkColor(board[yTemp][xTemp], oppColor)) break;
        }
    }
    return moves;
}

function queenMoves(board, color, xPos, yPos) {
    let moves = [];
    moves.push.apply(moves, bishopMoves(board, color, xPos, yPos));
    moves.push.apply(moves, rookMoves(board, color, xPos, yPos));
    return moves;
}

function kingMoves(board, color, xPos, yPos) {
    let moves = [];
    let oppColor = (color === 'w') ? 'b' : 'w';
    let xTemp;
    let yTemp;
    for (let i=0; i < kingDirections.length; i++) {
        xTemp = xPos;
        yTemp = yPos;
        if (xTemp >= 0 && xTemp < 8 && yTemp >= 0 && yTemp < 8) {
            xTemp += kingDirections[i][0];
            yTemp += kingDirections[i][1];
            if (!(xTemp >= 0 && xTemp < 8 && yTemp >= 0 && yTemp < 8)) continue;
            // stop if same color
            if (checkColor(board[yTemp][xTemp], color)) continue;
            moves.push([xTemp, yTemp]);
            // at first that can eat, continue
            if (checkColor(board[yTemp][xTemp], oppColor)) continue;
        }
    }
    return moves;
}

export function isInCheck(board, color, kingPos) {

    //let positions = color === "w" ? whitePositions : blackPositions;
    // let kingPos = [];
    // positions.forEach((value, key) => {
    //     if (value == 'k') kingPos = key;
    // });
    let xPos = kingPos[0];
    let yPos = kingPos[1];

    let oppColor = (color === 'w') ? 'b' : 'w';
    let _knightMoves = knightMoves(board, color, xPos, yPos);
    let _rookMoves   = rookMoves(board, color, xPos, yPos);
    let _bishopMoves = bishopMoves(board, color, xPos, yPos)
    let _kingMoves = kingMoves(board, color, xPos, yPos);
    let _pawnMoves;
    let _isInCheck = false;


    if (color == 'w') {
        _pawnMoves = [[xPos-1, yPos-1], [xPos+1, yPos-1]];
    } else {
        _pawnMoves = [[xPos-1, yPos+1], [xPos+1, yPos+1]];
    }

    _knightMoves.forEach(square => {
        if (checkColor(board[square[1]][square[0]], oppColor) && checkPiece(board[square[1]][square[0]], 'n')) {
            _isInCheck = true;
        }
     });

    if (_isInCheck) return true;

    _rookMoves.forEach(square => {
        if (checkColor(board[square[1]][square[0]], oppColor) && (checkPiece(board[square[1]][square[0]], 'r') || checkPiece(board[square[1]][square[0]], 'q'))) {
            _isInCheck = true;
        }
    });

    if (_isInCheck) return true;

    _bishopMoves.forEach(square => {
        if (checkColor(board[square[1]][square[0]], oppColor) && (checkPiece(board[square[1]][square[0]], 'b') || checkPiece(board[square[1]][square[0]], 'q'))) {
            _isInCheck = true;
        }
    });

    if (_isInCheck) return true;
    
    _kingMoves.forEach(square => {
        if (checkColor(board[square[1]][square[0]], oppColor) && checkPiece(board[square[1]][square[0]], 'k')) {
            _isInCheck = true;
        }
    });

    if (_isInCheck) return true;

    _pawnMoves.forEach(square => {
        if (square[0] >= 0 && square[0] < 8 && square[1] >= 0 && square[1] < 7
            && checkColor(board[square[1]][square[0]], oppColor) && checkPiece(board[square[1]][square[0]], 'p')) {
            _isInCheck = true;
        }
    });

    if (_isInCheck) return true;

    return false;
}

File: cpp/validMoves.cpp
#include "chess.h"


bool checkColor(const std::string& piece, char color) {
    if (piece.empty()) return false;
    return piece[0] == color;
}

bool checkPiece(const std::string& piece, char pieceType) {
    if (piece.empty()) return false;
    return piece[1] == pieceType;
}

void deleteFromMap(PositionsMap& positions, const std::pair<int, int>& key) {
    positions.erase(key);
}

bool isInCheck(std::string board[8][8], char color, const std::pair<int, int>& kingPos) {
    int xPos = kingPos.first;
    int yPos = kingPos.second;
    char oppColor = (color == 'w') ? 'b' : 'w';

    // Knight attacks
    std::vector<Move> knightAttacks;
    for (const auto& dir : knightDirections) {
        int x = xPos + dir.first;
        int y = yPos + dir.second;
        if (x >= 0 && x < 8 && y >= 0 && y < 8) {
            if (checkColor(board[y][x], oppColor) && checkPiece(board[y][x], 'n')) {
                return true;
            }
        }
    }

    // Rook and Queen attacks (horizontal and vertical)
    for (const auto& dir : rookDirections) {
        int x = xPos;
        int y = yPos;
        while (true) {
            x += dir.first;
            y += dir.second;
            if (x < 0 || x >= 8 || y < 0 || y >= 8) break;
            if (checkColor(board[y][x], color)) break;
            if (checkColor(board[y][x], oppColor)) {
                if (checkPiece(board[y][x], 'r') || checkPiece(board[y][x], 'q')) {
                    return true;
                } else {
                    break;
                }
            }
        }
    }

    // Bishop and Queen attacks (diagonals)
    for (const auto& dir : bishopDirections) {
        int x = xPos;
        int y = yPos;
        while (true) {
            x += dir.first;
            y += dir.second;
            if (x < 0 || x >= 8 || y < 0 || y >= 8) break;
            if (checkColor(board[y][x], color)) break;
            if (checkColor(board[y][x], oppColor)) {
                if (checkPiece(board[y][x], 'b') || checkPiece(board[y][x], 'q')) {
                    return true;
                } else {
                    break;
                }
            }
        }
    }

    // King attacks
    for (const auto& dir : kingDirections) {
        int x = xPos + dir.first;
        int y = yPos + dir.second;
        if (x >= 0 && x < 8 && y >= 0 && y < 8) {
            if (checkColor(board[y][x], oppColor) && checkPiece(board[y][x], 'k')) {
                return true;
            }
        }
    }

    // Pawn attacks
    int direction = (color == 'w') ? -1 : 1;
    for (int dx = -1; dx <= 1; dx += 2) {
        int x = xPos + dx;
        int y = yPos + direction;
        if (x >= 0 && x < 8 && y >= 0 && y < 8) {
            if (checkColor(board[y][x], oppColor) && checkPiece(board[y][x], 'p')) {
                return true;
            }
        }
    }

    return false;
}

std::vector<Move> validMovesFromArray(std::string board[8][8], int startX, int startY, const std::vector<Move>& movesArray, PositionsMap& positions, char color, GameData& g) {
    std::string pieceStr = board[startY][startX];
    // char color = pieceStr[0];
    char piece = pieceStr[1];
    std::pair<int, int> kingPos;
    for (const auto& [key, value] : positions) {
        if (value == 'k') {
            kingPos = key;
            break;
        }
    }

    std::vector<Move> validMoves;

    for (const auto& move : movesArray) {

        // handle en passant
        int flag = move.targetX == g.prevX && move.targetY == g.prevY && piece == 'p';
        int diffX = move.targetX - startX;

        int eatenY = move.targetY*(1-flag) + startY*flag;
        int eatenX = move.targetX*(1-flag) + (startX + diffX)*flag;

        std::string temp = board[move.targetY][move.targetX];
        std::string temp2 = board[eatenY][eatenX];
        board[eatenY][eatenX] = "";

        // Make the move
        board[startY][startX] = "";
        board[move.targetY][move.targetX] = pieceStr;

        std::pair<int, int> newKingPos = kingPos;
        if (piece == 'k') {
            newKingPos = {move.targetX, move.targetY};
        }

        if (!isInCheck(board, color, newKingPos)) {
            validMoves.push_back(move);
        }

        // Undo the move
        board[startY][startX] = std::string(1, color) + piece;
        board[move.targetY][move.targetX] = temp;
        board[eatenY][eatenX] = temp2;
    }

    return validMoves;
}

std::vector<Move> pawnMoves(std::string board[8][8], char color, int xPos, int yPos, GameData& g) {
    std::vector<Move> moves;
    bool isWhite = color == 'w';
    char oppColor; int direction; int startRow;
    
    if (isWhite) {
        oppColor = 'b';
        direction = -1;
        startRow = 6;
    } else {
        oppColor = 'w';
        direction = 1;
        startRow = 1;
    }
    
    int yNew = yPos+direction;
    // Single forward move
    if (yNew >= 0 && yNew < 8 && board[yNew][xPos].empty()) {
        moves.push_back({xPos, yPos, xPos, yNew});

        // if you can't move once you can't move twice
        yNew += direction;
        // it will be in bounds because its on a starting row
        if (yPos == startRow && board[yNew][xPos].empty()) {
            moves.push_back({xPos, yPos, xPos, yNew});
        }
    }

    // eating pieces normally
    int x = xPos - 1;
    int y = yPos + direction;
    if (y >= 0 && y < 8) {
        if (x >= 0 && x < 8 && checkColor(board[y][x], oppColor)) {
            moves.push_back({xPos, yPos, x, y});
        }
        x = xPos + 1;
        if (x >= 0 && x < 8 && checkColor(board[y][x], oppColor)) {
            moves.push_back({xPos, yPos, x, y});
        }
    }

    // eating pieces en passant
    // prevX, prevY are -2 if uninitialized so we need no check
    if (abs(g.prevX - xPos) == 1 && abs(g.prevY - yPos) == 1) {
        moves.push_back({xPos, yPos, g.prevX, g.prevY});
    }

    // Pawn promotion
    std::vector<Move> promotionMoves;
    for (auto& move : moves) {
        if (move.targetY == 0 || move.targetY == 7) {
            promotionMoves.push_back({move.startX, move.startY, move.targetX, move.targetY, 'q'});
            promotionMoves.push_back({move.startX, move.startY, move.targetX, move.targetY, 'r'});
            promotionMoves.push_back({move.startX, move.startY, move.targetX, move.targetY, 'b'});
            promotionMoves.push_back({move.startX, move.startY, move.targetX, move.targetY, 'n'});
        } else {
            promotionMoves.push_back(move);
        }
    }

    return promotionMoves;
}

std::vector<Move> rookMoves(std::string board[8][8], char color, int xPos, int yPos) {
    std::vector<Move> moves;
    char oppColor = (color == 'w') ? 'b' : 'w';

    for (const auto& dir : rookDirections) {
        int x = xPos;
        int y = yPos;
        while (true) {
            x += dir.first;
            y += dir.second;
            if (x < 0 || x >= 8 || y < 0 || y >= 8) break;
            if (checkColor(board[y][x], color)) break;
            moves.push_back({xPos, yPos, x, y});
            if (checkColor(board[y][x], oppColor)) break;
        }
    }
    return moves;
}

std::vector<Move> knightMoves(std::string board[8][8], char color, int xPos, int yPos) {
    std::vector<Move> moves;
    char oppColor = (color == 'w') ? 'b' : 'w';

    for (const auto& dir : knightDirections) {
        int x = xPos + dir.first;
        int y = yPos + dir.second;
        if (x >= 0 && x < 8 && y >= 0 && y < 8) {
            if (!checkColor(board[y][x], color)) {
                moves.push_back({xPos, yPos, x, y});
            }
        }
    }
    return moves;
}

std::vector<Move> bishopMoves(std::string board[8][8], char color, int xPos, int yPos) {
    std::vector<Move> moves;
    char oppColor = (color == 'w') ? 'b' : 'w';

    for (const auto& dir : bishopDirections) {
        int x = xPos;
        int y = yPos;
        while (true) {
            x += dir.first;
            y += dir.second;
            if (x < 0 || x >= 8 || y < 0 || y >= 8) break;
            if (checkColor(board[y][x], color)) break;
            moves.push_back({xPos, yPos, x, y});
            if (checkColor(board[y][x], oppColor)) break;
        }
    }
    return moves;
}

std::vector<Move> queenMoves(std::string board[8][8], char color, int xPos, int yPos) {
    std::vector<Move> moves = rookMoves(board, color, xPos, yPos);
    std::vector<Move> bishopMovesList = bishopMoves(board, color, xPos, yPos);
    moves.insert(moves.end(), bishopMovesList.begin(), bishopMovesList.end());
    return moves;
}

std::vector<Move> castleMoves(std::string board[8][8], char color, GameData& g) {
    std::vector<Move> moves;
    int yPos = (color == 'w') ? 7 : 0;
    bool hasKingMoved = (color == 'w') ? g.hasWhiteKingMoved : g.hasBlackKingMoved;
    if (hasKingMoved || isInCheck(board, color, {4, yPos})) return moves;

    if (color == 'w') {
        // Queenside castling
        if (!g.hasWhiteLeftRookMoved) {
            if (board[yPos][0] == "wr" && board[yPos][1].empty() && board[yPos][2].empty() && board[yPos][3].empty() && !isInCheck(board, color, {3, yPos})) {
                moves.push_back({4, yPos, 2, yPos});
            }
        }
        // Kingside castling
        if (!g.hasWhiteRightRookMoved) {
            if (board[yPos][7] == "wr" && board[yPos][5].empty() && board[yPos][6].empty() && !isInCheck(board, color, {5, yPos})) {
                moves.push_back({4, yPos, 6, yPos});
            }
        }
    } else {
        // Queenside castling
        if (!g.hasBlackLeftRookMoved) {
            if (board[yPos][0] == "br" && board[yPos][1].empty() && board[yPos][2].empty() && board[yPos][3].empty() && !isInCheck(board, color, {3, yPos})) {
                moves.push_back({4, yPos, 2, yPos});
            }
        }
        // Kingside castling
        if (!g.hasBlackRightRookMoved) {
            if (board[yPos][7] == "br" && board[yPos][5].empty() && board[yPos][6].empty() && !isInCheck(board, color, {5, yPos})) {
                moves.push_back({4, yPos, 6, yPos});
            }
        }
    }
    return moves;
}

std::vector<Move> kingMoves(std::string board[8][8], char color, int xPos, int yPos, GameData& g) {
    std::vector<Move> moves;
    char oppColor = (color == 'w') ? 'b' : 'w';

    // Normal king moves
    for (const auto& dir : kingDirections) {
        int x = xPos + dir.first;
        int y = yPos + dir.second;
        if (x >= 0 && x < 8 && y >= 0 && y < 8) {
            if (!checkColor(board[y][x], color)) {
                moves.push_back({xPos, yPos, x, y});
            }
        }
    }

    // Castling
    std::vector<Move> castlingMoves = castleMoves(board, color, g);
    moves.insert(moves.end(), castlingMoves.begin(), castlingMoves.end());

    return moves;
}

std::vector<Move> getValidMoves(std::string board[8][8], const std::string& piece, int xPos, int yPos, PositionsMap& positions, GameData& g) {
    char pieceType = piece[1];
    char color = piece[0];
    std::vector<Move> moves;

    switch (pieceType) {
        case 'p':
            moves = pawnMoves(board, color, xPos, yPos, g);
            break;
        case 'r':
            moves = rookMoves(board, color, xPos, yPos);
            break;
        case 'n':
            moves = knightMoves(board, color, xPos, yPos);
            break;
        case 'b':
            moves = bishopMoves(board, color, xPos, yPos);
            break;
        case 'q':
            moves = queenMoves(board, color, xPos, yPos);
            break;
        case 'k':
            moves = kingMoves(board, color, xPos, yPos, g);
            break;
    }

    // Filter moves that don't put the king in check
    moves = validMovesFromArray(board, xPos, yPos, moves, positions, color, g);

    return moves;
}


std::vector<Move> getValidMovesOfCurrentTurn(std::string board[8][8], PositionsMap& positions, GameData& g) {
    std::vector<Move> moves;
    for (const auto& [key, value] : positions) {
        int x = key.first;
        int y = key.second;
        std::string pieceStr = g.turn + std::string(1, value);
        std::vector<Move> pieceMoves = getValidMoves(board, pieceStr, x, y, positions, g);
        moves.insert(moves.end(), pieceMoves.begin(), pieceMoves.end());
    }
    return moves;
}